libdmmp/libdmmp.c

```diff
@@ -0,0 +1,437 @@
+			func_name, line);
+	} else {
+		fprintf(stderr, " # %s:%s():%d\n", file, func_name, line);
+	}
+}
+
+static int _init(struct dmmp_context *ctx)
+{
+	struct rlimit fd_limit;
+	int rc = DMMP_OK;
+	struct udev *udev = NULL;
+
+	udev = udev_new();
+	if (udev == NULL) {
+		rc = DMMP_ERR_NO_MEMORY;
+		_error(ctx, dmmp_strerror(rc));
+		return rc;
+	}
+
+	if (load_config(DEFAULT_CONFIGFILE, udev) != 0) {
+		rc = DMMP_ERR_LOAD_CONFIG_FAIL;
+		/* TODO(Gris Ge): Update load_config() to provide better error
+		 *		  message.
+		 */
+		_error(ctx, "%s: %s", dmmp_strerror(rc), DEFAULT_CONFIGFILE);
+		goto out;
+	}
+	_debug(ctx, "Config %s loaded", DEFAULT_CONFIGFILE);
+
+	if (conf->max_fds) {
+		/* Save current limit and restore it at _close() */
+		if (getrlimit(RLIMIT_NOFILE, &ctx->_rlimit) != 0) {
+			rc = DMMP_ERR_BUG;
+			_error(ctx, "Failed to get current "
+			       "RLIMIT_NOFILE limit: %d:%s", errno,
+			       strerror(errno));
+			goto out;
+		}
+
+		fd_limit.rlim_cur = conf->max_fds;
+		fd_limit.rlim_max = conf->max_fds;
+		if (setrlimit(RLIMIT_NOFILE, &fd_limit) != 0)
+			_warn(ctx, "can't set open fds limit to %d : %s",
+				conf->max_fds, strerror(errno));
+	}
+
+out:
+	if (udev != NULL)
+		udev_unref(udev);
+
+	if (rc != DMMP_OK)
+		_close(ctx);
+	return rc;
+}
+
+static void _close(struct dmmp_context *ctx)
+{
+	dm_lib_release();
+	dm_lib_exit();
+
+	cleanup_prio();
+	cleanup_checkers();
+
+	free_config(conf);
+	conf = NULL;
+
+	if (ctx->_has_rlimit == true) {
+		if (setrlimit(RLIMIT_NOFILE, &ctx->_rlimit) != 0)
+			_warn(ctx, "_close(): failed to restore rlimit: %d: %s",
+			      errno, strerror(errno));
+	}
+}
+
+/*
+ * Copy from multipath/main.c update_paths(). Just changed the return to void.
+ */
+static void _update_path_status(struct multipath *mpp)
+{
+	int i, j;
+	struct pathgroup * pgp;
+	struct path * pp;
+
+	if (!mpp->pg)
+		return;
+
+	vector_foreach_slot (mpp->pg, pgp, i) {
+		if (!pgp->paths)
+			continue;
+
+		vector_foreach_slot (pgp->paths, pp, j) {
+			if (!strlen(pp->dev)) {
+				if (devt2devname(pp->dev, FILE_NAME_SIZE,
+						 pp->dev_t)) {
+					/*
+					 * path is not in sysfs anymore
+					 */
+					pp->chkrstate = pp->state = PATH_DOWN;
+					continue;
+				}
+				pp->mpp = mpp;
+				if (pathinfo(pp, conf->hwtable, DI_ALL))
+					pp->state = PATH_UNCHECKED;
+				continue;
+			}
+			pp->mpp = mpp;
+			if (pp->state == PATH_UNCHECKED ||
+			    pp->state == PATH_WILD) {
+				if (pathinfo(pp, conf->hwtable, DI_CHECKER))
+					pp->state = PATH_UNCHECKED;
+			}
+
+			if (pp->priority == PRIO_UNDEF) {
+				if (pathinfo(pp, conf->hwtable, DI_PRIO))
+					pp->priority = PRIO_UNDEF;
+			}
+		}
+	}
+	return;
+}
+
+static int _get_pathvec(struct dmmp_context *ctx, vector *pathvec)
+{
+	int rc = DMMP_OK;
+	int di_flag = DI_SYSFS | DI_CHECKER;
+	/* ^ like multipath -ll to check path status */
+	int rc_discovery = 0;
+
+	*pathvec = vector_alloc();
+	/* No need to whether pathvec is NULL, caller already done */
```

Comment:
The `_dmmp_alloc_null_check(ctx, *pathvec, rc, out);` will check whether `*pathvec` is NULL or not.